### PR TITLE
Add Mined Block list in account page

### PR DESCRIPTION
**Pull request by x86chi**
## Preview

Screen Shot 2020-08-10 at 11 20 30 AM

### CODE REVIEW

**Comment by x86chi** [comment]
I've Update GraphQL Schema and API. But it still response to 500 code.

**Comment by x86chi** [comment]
## Preview

Screen Shot 2020-08-10 at 11 20 30 AM

## TODO

- [ ] Add Pagenation UI

File: src/subpages/list.tsx [hunk 1]
```diff
@@ -1,259 +1,133 @@
...
+            maxWidth: 200,
+            ...commonProps,
+            isSortedDescending: true,
+            data: 'string',
+            isPadded: true,
+            onRender: (block: Block, index: number) => {
+              if (blocks === null) throw Error('blocks is null');
+              // FIX TODO: beforeBlock index is not available
+              const beforeBlock =
+                blocks[Math.min(index! + 1, block.length - 1)];
```
Comment (x86chi) [inline src/subpages/list.tsx:52]: Time taken cell couldn't calculating.

File: src/subpages/list.tsx [hunk 2]
```diff
@@ -1,259 +1,133 @@
...
-                    : (data!.blockQuery!.blocks as NonNullable<Block[]>)
-                }
-                loading={loading}
+              <SummaryCards blocks={blocks} />
+              <OffsetSwitch
+                olderHandler={olderHandler}
+                newerHandler={newerHandler}
+                disable={{ older: loading || offset < 1, newer: loading }}
               />
+              <BlockList blocks={blocks} loading={loading} columns={columns} />
```
Comment (x86chi) [inline src/subpages/list.tsx]: So I excluded the `timeTaken` column.

Comment (x86chi) [inline src/subpages/list.tsx:66]: So I excluded the `timeTaken` column. instead of below line.

```suggestion
              <BlockList blocks={blocks} loading={loading} columns={[...columns, timeTaken]]} />
```

**Comment by x86chi** [comment]
Screen Shot 2020-08-10 at 7 40 28 PM

## TODO

- [ ] split query string

File: gatsby-node.js [hunk 3]
```diff
@@ -27,14 +27,20 @@ exports.createPages = async ({ graphql, actions }) => {
...
           endpoint,
         },
-      });
+      };
+
+      if (file === 'list.tsx' || file === 'search.tsx') {
+        createPage(pageProps);
+        return;
+      }
+      createPage({ ...pageProps, matchPath: pagePath + ':id' });
```
Comment (x86chi) [inline gatsby-node.js:43]: search page can't found.

`account`, `block`, `transaction` subpage use `:id` path.

Comment (dahlia) [reply gatsby-node.js:43]: This mechanism requires server-side rendering and without this links will work only at client-side, i.e., only available through links, not direct access to URLs.  Here's the repro:

```bash
./build.sh
cd _site/
npx http-server 8989
```

And then try to open <http://127.0.0.1:8989/9c-beta/block/fe41ed19f20ae643828d9fea514085a6069ce0a7c2752e19e3939a2d89000000> on the web browser:

404 Not Found

The page won't be loaded, since the server could not find the file like *9c-beta/block/fe41ed19f20ae643828d9fea514085a6069ce0a7c2752e19e3939a2d89000000*.

Comment (x86chi) [reply gatsby-node.js:43]: Ok I'll revert change.

File: package.json [hunk 4]
```diff
@@ -12,7 +12,8 @@
     "dev:gatsby": "gatsby develop -H 0.0.0.0",
     "dev:prettier": "onchange \"@(*.css|*.js|*.json|*.ts|@(src|types)/**/*.@(ts|tsx))\" -- prettier --write {{changed}}",
     "codegen": "graphql-codegen --config codegen.yml",
-    "dev": "run-p dev:**"
+    "dev": "run-p dev:**",
+    "tsc": "tsc --noemit"
```
Comment (moreal) [inline package.json:16]:
```suggestion
    "tsc": "tsc --noEmit"
```

It also works but `--noEmit` seems right representation by following [compiler-options document][typescript-compiler-options-docs].

[typescript-compiler-options-docs]: https://www.typescriptlang.org/docs/handbook/compiler-options.html

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Use Node.js
+        uses: actions/setup-node@v1
+        with:
+          node-version: 10.x
+      - name: Install dependencies
+        run: npm i
+      - name: Type Check
+        run: npm run tsc
```
Comment (moreal) [inline .github/workflows/ci.yml:16]: Actually, IMHO it should be handle as another pull request because there is no relation to its feature, *Add Mined Block list in account page*.

File: src/misc/useOffset.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import useSearchParams from './useSearchParams';
+
+export const limit = 21;
+export default function useOffset(location: Location, keyName = 'offset') {
+  const [searchParams, setSearchParams] = useSearchParams(location);
+  const offset = keyName in searchParams ? searchParams[keyName] : 0;
+  const setOffset = (offset: number) => {
+    if (offset < 1) {
+      const newSearchParams = { ...searchParams };
+      delete newSearchParams[keyName];
```
Comment (moreal) [inline src/misc/useOffset.ts:12]: I wonder why it deletes after copy `searchParams`, not delete directly from `searchParams`.

File: src/subpages/list.tsx [hunk 7]
```diff
@@ -1,263 +1,138 @@
...
+              if (blocks === null || index === undefined)
+                throw Error('blocks is null');
+              // FIX TODO: beforeBlock index is not available
+              const beforeBlock =
+                blocks[Math.min(index + 1, blocks.length - 1)];
+              const beforeTimestamp = Date.parse(beforeBlock.timestamp);
+              const nowTimestamp = Date.parse(block.timestamp);
+              return <>{(nowTimestamp - beforeTimestamp) / 1000}</>;
+            },
+          };
```
Comment (moreal) [inline src/subpages/list.tsx:58]: `timeTaken` seems not used anywhere, does it?

File: src/subpages/list.tsx [hunk 8]
```diff
@@ -1,263 +1,138 @@
...
+            name: 'Time Taken',
+            minWidth: 50,
+            maxWidth: 200,
+            ...commonProps,
+            isSortedDescending: true,
+            data: 'string',
+            isPadded: true,
+            onRender: (block: Block, index) => {
+              if (blocks === null || index === undefined)
+                throw Error('blocks is null');
```
Comment (moreal) [inline src/subpages/list.tsx:50]: I wonder the case when `index === undefined` became true.

File: src/subpages/list.tsx [hunk 9]
```diff
@@ -1,263 +1,138 @@
...
+  const timestamps: Date[] = blocks.map(block => new Date(block.timestamp));
+
+  let interval = 0;
+  for (let i = 0; i < timestamps.length - 1; i++) {
+    interval += +timestamps[i] - +timestamps[i + 1];
+  }
+  interval /= (timestamps.length - 1) * 1000;
+
+  const difficulties = blocks.map(block => block.difficulty);
+  const difficulty =
```
Comment (moreal) [inline src/subpages/list.tsx:99]: I believe It will be better to use another variable name like `difficultyAverage` or better one because it means *average of difficulties of blocks*. Also `interval` too.

File: src/subpages/list.tsx (hunk 8 above)
Comment (x86chi) [reply src/subpages/list.tsx:50]: It just type guard. index can be undefined.

Screen Shot 2020-08-12 at 4 50 44 PM

File: src/subpages/list.tsx (hunk 7 above)
Comment (x86chi) [reply src/subpages/list.tsx:58]: transaction list render dons't work. so I add this for comment.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (x86chi) [reply .github/workflows/ci.yml:16]: It was difficult to type check, so I #102 closed and merged to this branch.

Comment (moreal) [reply .github/workflows/ci.yml:16]: Yes, I am strongly sympathetic and agree with your opinion, difficulty to check types. But what I wanted to say was, what relationships exist between creating GitHub Actions Workflow and adding *Mined Block list in account page*.

File: src/subpages/list.tsx (hunk 8 above)
Comment (moreal) [reply src/subpages/list.tsx:50]: Then the error message seems not to suit to this case because it will be always `'blocks is null'` even if index is `undefined`.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (x86chi) [reply .github/workflows/ci.yml:16]: Ok I 'll remove this file.

**Comment by limebell** [comment]
Some files conflicts 😥  Could you please rebase this on main branch?

File: src/subpages/list.tsx (hunk 9 above)
Comment (x86chi) [reply src/subpages/list.tsx:99]: I agree rename to difficultyAverage. But I think `interval` is more readable.

**Comment by x86chi** [comment]
> Some files conflicts 😥 Could you please rebase this on main branch or resolve it?

I did a rebase.

**Comment by moreal** [comment]
@x86chi It seems you rebased even commits you didn't. Could you check it again, please?

image

**Comment by x86chi** [comment]
Finished rebase. 😓

**Comment by x86chi** [comment]
These a inteded result. But it would be better to squash or delete commits?

**Comment by x86chi** [comment]
I've squashed to 4 commit.

File: src/subpages/list.tsx [hunk 10]
```diff
@@ -1,259 +1,134 @@
...
+            ...commonProps,
+            isSortedDescending: true,
+            data: 'string',
+            isPadded: true,
+            onRender: (block: Block, index) => {
+              if (blocks === null) throw Error('blocks is null');
+              if (index === undefined) throw Error('index is null');
+              // FIX TODO: beforeBlock index is not available
+              const beforeBlock =
+                blocks[Math.min(index + 1, blocks.length - 1)];
```
Comment (x86chi) [inline src/subpages/list.tsx:53]: FIX TODO : beforeBlock indexing not work.

Comment (x86chi) [reply src/subpages/list.tsx]: I've sovled this issue in 78db502.

It is calculated only inside the component wrapped.

**Comment by x86chi** [comment]
> This pull request seems to have more than 30 commits… is it intended @x86chi?

@dahlia Please to approve from request change.

File: src/subpages/list.tsx [hunk 11]
```diff
@@ -132,3 +114,34 @@ const Cards: React.FC<CardsProps> = ({
...
+    minWidth: 50,
+    maxWidth: 200,
+    ...commonProps,
+    isSortedDescending: true,
+    data: 'string',
+    isPadded: true,
+    onRender: (block: Block, index) => {
+      if (blocks === null) throw Error('blocks is null');
+      if (index === undefined) throw Error('index is null');
+      const beforeBlock = blocks[Math.min(index + 1, blocks.length - 1)];
```
Comment (x86chi) [inline src/subpages/list.tsx:129]:
```suggestion
     const beforeBlock = blocks.find(({ index }) => index === block.index - 1);
     if (beforeBlock === undefined) return 0;
```

How about this for index accuracy?

Comment (dahlia) [reply src/subpages/list.tsx:129]: Both seem to behave in the same way, except for time complexity.  Whereas index operator is random access, `.find()` is O(n) at least.

Comment (dahlia) [reply src/subpages/list.tsx:129]: Well, if we can assume `blocks` are always short I believe @x86chi's suggestion would be good.
